src/network/networkctl.c

```diff
@@ -2919,6 +2930,488 @@ static int verb_reconfigure(int argc, char *argv[], void *userdata) {
+
+        *ret_path = TAKE_PTR(p);
+        *ret_dropins = TAKE_PTR(dropins_split);
+
+        return 0;
+}
+
+static int parse_link_config(
+                const char *link_config,
+                char **ret_ifname,
+                char **ret_type) {
+
+        _cleanup_strv_free_ char **split = NULL;
+        _cleanup_free_ char *ifname = NULL, *type = NULL;
+        size_t n;
+
+        assert(link_config);
+        assert(ret_ifname);
+        assert(ret_type);
+
+        split = strv_split(link_config, ":");
+        if (!split)
+                return log_oom();
+
+        n = strv_length(split);
+        if (n == 0 || isempty(split[0]))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No link name is given.");
+        if (n > 2)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid link config '%s'.", link_config);
+
+        ifname = strdup(split[0]);
+        type = strdup(n == 2 ? split[1] : "network");
+        if (!ifname || !type)
+                return log_oom();
+
+        *ret_ifname = TAKE_PTR(ifname);
+        *ret_type = TAKE_PTR(type);
+
+        return 0;
+}
+
+static int add_config_to_edit(
+                EditFileContext *context,
+                const char *path,
+                char * const *dropins) {
+
+        _cleanup_free_ char *new_path = NULL, *dropin_path = NULL, *old_dropin = NULL;
+        _cleanup_strv_free_ char **comment_paths = NULL;
+        int r;
+
+        assert(context);
+        assert(path);
+
+        if (path_startswith(path, "/usr")) {
+                _cleanup_free_ char *name = NULL;
+
+                r = path_extract_filename(path, &name);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to extract filename from '%s': %m", path);
+
+                new_path = path_join(NETWORK_DIRS[0], name);
+                if (!new_path)
+                        return log_oom();
+        }
+
+        if (!arg_drop_in)
+                return edit_files_add(context, new_path ?: path, path, NULL);
+
+        assert(dropins);
+
+        r = get_dropin_by_name(arg_drop_in, dropins, &old_dropin);
+        if (r < 0)
+                return log_error_errno(r, "Failed to acquire drop-in '%s': %m", arg_drop_in);
+
+        if (r > 0 && !path_startswith(old_dropin, "/usr"))
+                /* An existing drop-in is found and not in /usr/. Let's edit it directly. */
+                dropin_path = TAKE_PTR(old_dropin);
+        else {
+                /* No drop-in was found or an existing drop-in resides in /usr/. Let's create
+                 * a new drop-in file. */
+                _cleanup_free_ char *dropin_dir = NULL;
+
+                dropin_dir = strjoin(new_path ?: path, ".d");
+                if (!dropin_dir)
+                        return log_oom();
+
+                dropin_path = path_join(dropin_dir, arg_drop_in);
+                if (!dropin_path)
+                        return log_oom();
```

Comment:
Not a matter, but this can be shorten by `strjoin(new_path ?: path, ".d/", arg_drop_in)`.